validators/types_unix.go

```diff
@@ -56,7 +56,10 @@ var DefaultSysSpec = SysSpec{
 		},
 		Forbidden: []KernelConfig{},
 	},
-	Cgroups: []string{"cpu", "cpuacct", "cpuset", "devices", "freezer", "memory"},
+	CgroupSpec: CgroupSpec{
+		Required: []string{"cpu", "cpuacct", "cpuset", "devices", "freezer", "memory"},
+		Optional: []string{"hugetlb", "pids"},
```

Comment:
@dashpole IIRC there was a reason other than it not existing on all systems to not check for pids here, or am I remembering it incorrectly?

Reply:
Could you please document the reason why these are optional on certain versions?

Reply:
> Could you please document the reason why these are optional on certain versions?

Sure! 

### hugetlb
The `hugetlb` cgroup is used for enforcing the limits on pods using [huge pages](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190129-hugepages.md#cgroup-enforcement). The current kubelet implementation (all versions with huge page support) just ignore it if the cgroup is missing, aka. not enforcing the limits. Some kernels are compiled without huge page support (resulting in the hugetlb cgroup to be missing). It is also possible to compile the kernel without the hugetlb cgroup but with huge page support, but I see no real reason why one should decide to to that).

The hugetlb cgroup is from 2012, so all "modern" kernels should support it if they support huge pages and cgroups. Setting it as "required" would then make sense, but since there might be reasons to compile the kernel without huge page support, i think keeping it optional is ok, or should it be required when huge pages are supported? If the hugetlb cgroup is not enabled on a node while huge pages are supported, the node will silently ignore it and _skip_ enforcing it (but that is another discussion).

Huge page support has been in `GA` since `v1.14`.


### pids
The `pids` cgroup is used for enforcing [pid limits](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190129-pid-limiting.md). There are two feature flags, `SupportPodPidsLimit` & `SupportNodePidsLimit`. Both are in beta (from `v1.14` and `v1.15`), and therefore enabled by default. The current implementation will require the `pids` cgroup as long as one of them is enabled, and since both are enabled by default, the cgroup is essentially required.

The `pids` cgroup was introduced in 2015, so it is a bit newer than `hugetlb`. @RobertKrawitz and @dashpole probably have some more insights.

Reply:
Yeah, unless the validator has access to feature gates, it seems reasonable to mark pids as optional.

Reply:
@odinuge could you put a brief summary of this in the code so that when someone looks at it they'll understand the rationale?

Reply:
> Could you please document the reason why these are optional on certain versions?

i think having some comments in the code make sense too.
possibly not the full explanation, but a short summary @odinuge

Reply:
+1 to code comment.